src/python/pants/backend/jvm/rules/targets.py

```diff
@@ -0,0 +1,808 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from typing import Any, Dict, Iterable, Optional, Tuple, Union
+
+from pants.backend.jvm.targets.jvm_binary import JarRules
+from pants.build_graph.address import Address
+from pants.engine.target import (
+    COMMON_TARGET_FIELDS,
+    BoolField,
+    Dependencies,
+    DictStringToStringField,
+    DictStringToStringSequenceField,
+    IntField,
+    InvalidFieldTypeException,
+    PrimitiveField,
+    ProvidesField,
+    Sources,
+    StringField,
+    StringSequenceField,
+    Target,
+)
+from pants.java.jar.exclude import Exclude
+from pants.java.jar.jar_dependency import JarDependency
+from pants.util.collections import ensure_list, ensure_str_list
+from pants.util.frozendict import FrozenDict
+
+# -----------------------------------------------------------------------------------------------
+# Common JVM Fields
+# -----------------------------------------------------------------------------------------------
+
+
+class JvmExcludes(PrimitiveField):
```

Comment:
Either pre- or post-work of this PR, it would be nice to introduce `TupleField` and `DictField` :)

Reply:
We have `StringSequenceField` and `DictStringToStringField` and `DictStringToStringSequenceField`. 

We don't have a generic `SequenceField` yet because I haven't figured out a good way to abstract it in a way that works with type hints still, but I think I can follow up with this in a way that at least removes most duplication of the validation. Good idea.

We'll probably never have `DictField`. The validation + logic to convert to `FrozenDict` (we must ensure everything is hashable) is far too complex to generalize.